android/KMEA/app/src/main/java/com/tavultesoft/kmea/ModelPickerActivity.java

```diff
@@ -75,115 +81,173 @@ public void onCreate(Bundle savedInstanceState) {
-            // we only need to add the model to the list of installed models
-            HashMap<String, String> modelInfo = new HashMap<String, String>();
-            modelInfo.put(KMManager.KMKey_PackageID, packageID);
-            modelInfo.put(KMManager.KMKey_LexicalModelID, modelID);
-            modelInfo.put(KMManager.KMKey_LanguageID, languageID);
-            modelInfo.put(KMManager.KMKey_LexicalModelName, modelName);
-            modelInfo.put(KMManager.KMKey_LanguageName, langName);
-            modelInfo.put(KMManager.KMKey_LexicalModelVersion, version);
-            // Add help link
-            modelInfo.put(KMManager.KMKey_CustomHelpLink, "");
-
-            boolean result = KMManager.addLexicalModel(context, new HashMap<>(modelInfo));
-            if (result) {
-              Toast.makeText(context, getString(R.string.model_install_toast), Toast.LENGTH_SHORT).show();
-            }
-
-            immediateRegister = true;
-          } else {
-            // Model isn't installed so prompt to download it
-            Bundle args = model.buildDownloadBundle();
-            Intent i = new Intent(getApplicationContext(), KMKeyboardDownloaderActivity.class);
-            i.putExtras(args);
-            startActivity(i);
-          }
+    // Establish the list view based on the CloudRepository's Dataset
+    repo = CloudRepository.shared.fetchDataset(context);
+
+    // add listener to dataset to get event for catalog update.
+    repoObserver = new DataSetObserver() {
+      @Override public void onChanged() {
+        updateProgressBar();
+      }
+    };
+    repo.registerDataSetObserver(repoObserver);
+
+    // Check if we need to query for associated dictionary
+    LexicalModel lm = CloudRepository.shared.getAssociatedLexicalModel(context, newLanguageID);
+    if (lm == null) {
+      if (KMManager.hasConnection(context)) {
+        // Query to install associated lexical model
+        String _downloadid = CloudLexicalModelMetaDataDownloadCallback.createDownloadId(newLanguageID);
+        CloudLexicalModelMetaDataDownloadCallback _callback = new CloudLexicalModelMetaDataDownloadCallback();
+
+        Toast.makeText(context,
+          context.getString(R.string.query_associated_model),
+          Toast.LENGTH_SHORT).show();
+
+        ArrayList<CloudApiTypes.CloudApiParam> aPreparedCloudApiParams = new ArrayList<>();
+        String url = CloudRepository.prepareLexicalModelQuery(newLanguageID);
+        aPreparedCloudApiParams.add(new CloudApiTypes.CloudApiParam(
+          CloudApiTypes.ApiTarget.KeyboardLexicalModels, url).setType(CloudApiTypes.JSONType.Array));
+
+        CloudDownloadMgr.getInstance().executeAsDownload(
+          context, _downloadid, null, _callback,
+          aPreparedCloudApiParams.toArray(new CloudApiTypes.CloudApiParam[0]));
+      } else {
+        Toast.makeText(context,
+          "Network access needed to check for available dictionary",
```

Comment:
i18n: string should be localizable.

Reply:
Replaced with a localizable string "Cannot connect to Keyman server!"